### PR TITLE
Handle tasty files reported twice

### LINKED ISSUE #964: [sbt 1.5.0-M2] Zinc crashes when compiling the Test config with Scala 3
## steps

sbt.version= 1.5.0-M2
scalaVersion := "3.0.0-RC1"

1. Create a Scala 3 file in `src/test`
```scala
// src/test/scala/foo/Foo.scala
package foo

def foo(a: Any): Unit = ???
```
2. Run `Test / compile`
3. Update the `Foo.scala` file
4. Run `Test / compile`

```shell
sbt:root> Test / compile
[error] ## Exception when compiling 1 sources to /home/piquerez/scalacenter/tasty-macros/target/scala-3.0.0-RC1/test-classes
[error] java.lang.IllegalArgumentException: requirement failed: Source file '/home/piquerez/scalacenter/tasty-foo/target/scala-3.0.0-RC1/test-classes/foo/Foo$package.tasty' does not exist.
[error] scala.Predef$.require(Predef.scala:281)
[error] sbt.io.IO$.copyFile(IO.scala:895)
[error] sbt.io.IO$.move(IO.scala:1118)
[error] sbt.internal.inc.ClassFileManager$TransactionalClassFileManager.move(ClassFileManager.scala:241)
[error] sbt.internal.inc.ClassFileManager$TransactionalClassFileManager.$anonfun$deleteImpl$6(ClassFileManager.scala:200)
[error] scala.collection.Iterator.foreach(Iterator.scala:943)
[error] scala.collection.Iterator.foreach$(Iterator.scala:943)
[error] scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[error] scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error] scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error] scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error] sbt.internal.inc.ClassFileManager$TransactionalClassFileManager.deleteImpl(ClassFileManager.scala:200)
[error] sbt.internal.inc.ClassFileManager$TransactionalClassFileManager.delete(ClassFileManager.scala:191)
[error] xsbti.compile.WrappedClassFileManager.delete(WrappedClassFileManager.java:45)
[error] sbt.internal.inc.IncrementalCommon$.pruneClassFilesOfInvalidations(IncrementalCommon.scala:840)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:93)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:587)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2301)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2258)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2254)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] java.base/java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by adpi2**
fix #964 

An auxiliary class file (`.tasty`, `.sjsir`...) can be reported by 2 different instances of `AuxiliaryClassFiles`. It probably means Zinc is poorly configured but it is also safer to handle these cases in Zinc by calling `.distinct` on the class files and auxiliary files.

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/965/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

**Comment by smarter** [comment]
Side-note, but I'd be in favor of enabling the .tasty/.sjsir/.nir file manager handling by default, so that we don't have to worry about where it should be enabled in every build tool, I don't think it would break anything since we have to handle the fact that the file we're trying to delete doesn't exist anyway.

**Comment by adpi2** [comment]
That's probably a good idea.

**Comment by adpi2** [comment]
That's probably a good idea. But also we should keep the configuration open in case some compiler plugins want to emit new files.

File: internal/zinc-core/src/main/scala/sbt/internal/inc/ClassFileManager.scala [hunk 1]
```diff
@@ -22,6 +22,22 @@ import java.nio.file.{ Files, Path }
 import java.util.Optional
 import scala.collection.mutable
 
+private abstract class AbstractClassFileManager(auxiliaryFiles: Iterable[AuxiliaryClassFiles])
+    extends XClassFileManager {
+  protected def allPaths(classes: Vector[Path]): Vector[Path] = {
+    // different auxiliaryFiles can produce the same associated file
+    // we remove the duplicates by calling distinct
+    (classes ++ classes.flatMap(associatedFiles)).distinct
```
Comment (smarter) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/ClassFileManager.scala]: Maybe not a big deal, I worry a bit about the efficiency of `distinct` if you have a really big project. I'm not sure this is full-proof either (could you end up with a relative and absolute path to the same file that would be considered distinct?), so I'd be OK with an alternative fix where we just change `TransactionalClassFileManager#deleteImpl` to check that the file exists before calling `move` on it, your call.

Comment (adpi2) [reply internal/zinc-core/src/main/scala/sbt/internal/inc/ClassFileManager.scala]: As you suggest I removed the `.distinct` and moved the check that the files exists just before the call to `move`. Other implementations of `ClassFileManager` seems to be fine already.

**Review by smarter — APPROVED** [review]

**Review by eed3si9n — APPROVED** [review]

**Pull request merged**